### DISCUSSION TITLE
What is Native and how is it used?

### DISCUSSION THREAD

**Discussion by DamianReeves** [finos/morphir#56] [Q&A] [answered]
I've noticed morphir-elm now includes something called `Native` could we get a description of the feature, and how one would use it?

**Comment by AttilaMihaly** [comment, marked as answer]
Very timely question. @piyush-gupta1999, our most productive contributor just told me yesterday that we should document this. I'll give a high-level description here and will add code comments as well.

Functions in the Native module are called by the interpreter when it hits an SDK function. Things like adding or comparing two numbers cannot be implemented using the Morphir IR itself so we call the built-in SDK functions but we are working with `Value`s in the interpreter so we have to unwrap, get the literal value out. invoke the SDK function and wrap the result back to a `Value` before returning. The Native module contains types that describe the interface for native functions and also helpers to simplify the unwrap/wrap process.

Does that answer your question?

**Reply by DamianReeves** [reply]
So its safe to say that Native is more about the Elm tooling?

**Reply by AttilaMihaly** [reply]
Yes. More specifically I would say it's about the Morphir interpreter which is currently only available in the Elm tooling. If you wanted to port the interpreter to Scala for example you would have the choice of using reflection or implementing a similar Native module to invoke the Scala SDK for built-in calculations.
